common/djangoapps/entitlements/models.py

```diff
@@ -24,3 +54,65 @@ class CourseEntitlement(TimeStampedModel):
         help_text='The current Course enrollment for this entitlement. If NULL the Learner has not enrolled.'
     )
     order_number = models.CharField(max_length=128, null=True)
+    _policy = models.ForeignKey(CourseEntitlementPolicy, null=True, blank=True)
+
+    @property
+    def expired_at_datetime(self):
+        if not self.expired_at and self.get_days_until_expiration() < 0:
+            self.expired_at = datetime.utcnow()
+            self.save()
+        return self.expired_at
+
+    @expired_at_datetime.setter
+    def expired_at_datetime(self, value):
+        self.expired_at = value
+
+    @property
+    def policy(self):
+        return self._policy or CourseEntitlementPolicy()
+
+    @policy.setter
+    def policy(self, value):
+        self._policy = value
+
+    def get_days_since_created(self):
+        """
+        Returns an integer of number of days since the entitlement has been created
+        """
+        utc = pytz.UTC
+        return (datetime.now(tz=utc) - self.created).days
+
+    def get_days_until_expiration(self):
+        """
+        Returns an integer of number of days until the entitlement expires
+        """
+        expiry_date = self.created + timedelta(self.policy.expiration_period_days)
+        now = datetime.now(tz=pytz.UTC)
+        return (expiry_date - now).days
+
+    def is_entitlement_regainable(self):
```

Comment:
In general I think that we should be asking the policy these questions, not the entitlement.
`return policy.is_entitlement_regainable(self)`

Applying a policy should mean using a common policy API to determine information about an entitlement.  A policy should be more than just a key value store of different date limits.

This will prevent this method from getting bloated in the future when it needs to handle different regain cases. 
```
if policy.type=="golden_ticket":
    return True
elif self.enrollment_course_run:
   # do stuff
elif _has_active_prerequisites(self.course_uuid):
   # handle prerequisite case
```

Reply:
This was not meant to be a generic policy manager.  Maybe we named it incorrectly.  If we need something more robust in the future we should tackle that.